### DISCUSSION TITLE
batch_dice for 3d_fullres depends on whether a 3d_cascade_fullres was planned as well

### DISCUSSION THREAD

**Discussion by BennoGesierich** [MIC-DKFZ/nnUNet#1301] [Q&A] [answered]
Hi, is it correct that the batch_dice is set differently for `3d_fullres` networks, depending on whether a cascade was triggered during planning or not?
Seems like this distinction is done in script:
https://github.com/MIC-DKFZ/nnUNet/blob/master/nnunet/run/default_configuration.py
in line 71ff:
```
if (network == '2d' or len(possible_stages) > 1) and not network == '3d_lowres':
        batch_dice = True
        print("I am using batch dice + CE loss")
    else:
        batch_dice = False
        print("I am using sample dice + CE loss")
```

I was wondering why batch_dice should be set differently, if 3d_fullres is trained (and not 3d_cascade_fullres) in the context of large images, which trigger the cascade during planning? Would this cause different training results for 3d_fullres, compared to a training where planning the cascade is suppressed?

**Comment by FabianIsensee** [comment, marked as answer]
Yes it's intentional.
If there exists no 3d_lowres then we can assume that the patch size is similar to the individual image size, so computing the dice loss per sample in the batch makes sense as each sample is essentially one image. This assumes, of course, that the target metric is the dice score, computed for each image independently and then averaged over all images.
Conversely, if a 3d_lowres stage exists, then patch_size << image_size. Thus, we know each patch only shows a fraction of an entire image and the class distributions in the patches most likely do not reflect the true class distribution in the images. This can cause stability problems. Pretending all samples/patches in the batch are just one image solves that

**Reply by BennoGesierich** [reply]
Thank you very much, I understand it now. 
I changed in my custom planner class the parameter `self.how_much_of_a_patient_must_the_network_see_at_stage0 = 1` to trigger the cascade also in cases, where the patch size is more close to the actual image size than 1/4. I guess that wasn't necessarily a good idea then, given this side effect. Or, maybe, I should at least plan the 3d_fullres experiment in this case additionally/separately with the standard planner class.
